File: ticket-service/src/main/java/com/eventbooking/ticket/service/TicketService.java
package com.eventbooking.ticket.service;

import com.eventbooking.ticket.dto.GenerateTicketsRequest;
import com.eventbooking.ticket.dto.TicketDto;

import java.util.List;
import java.util.UUID;

public interface TicketService {
    
    /**
     * Generate tickets for a completed order
     * @param request The ticket generation request
     * @return List of generated tickets
     */
    List<TicketDto> generateTickets(GenerateTicketsRequest request);
    
    /**
     * Get a ticket by its ID
     * @param ticketId The ticket ID
     * @return The ticket DTO
     */
    TicketDto getTicketById(UUID ticketId);
    
    /**
     * Get a ticket by its ticket number
     * @param ticketNumber The ticket number
     * @return The ticket DTO
     */
    TicketDto getTicketByNumber(String ticketNumber);
    
    /**
     * Get all tickets for an order
     * @param orderId The order ID
     * @return List of tickets
     */
    List<TicketDto> getTicketsByOrderId(UUID orderId);
    
    /**
     * Get all tickets for a user (by retrieving all their orders)
     * @param userId The user ID
     * @return List of tickets
     */
    List<TicketDto> getTicketsByUserId(UUID userId);
    
    /**
     * Cancel a ticket
     * @param ticketId The ticket ID
     */
    void cancelTicket(UUID ticketId);
    
    /**
     * Validate a ticket by QR code
     * @param qrCode The QR code data
     * @return The ticket DTO if valid
     */
    TicketDto validateTicket(String qrCode);
}

File: frontend/src/services/ticketService.ts
import { apiRequest } from './api';

export enum TicketStatus {
  ACTIVE = 'ACTIVE',
  USED = 'USED',
  CANCELLED = 'CANCELLED',
  EXPIRED = 'EXPIRED',
}

export interface Ticket {
  id: string;
  ticketTypeId: string;
  orderId: string;
  ticketNumber: string;
  qrCode: string;
  holderName: string;
  status: TicketStatus;
  createdAt: string;
  updatedAt: string;
  // Event and ticket type details
  eventName?: string;
  eventDate?: string;
  venueName?: string;
  venueAddress?: string;
  ticketTypeName?: string;
  venueZone?: string;
}

export const ticketService = {
  // Get ticket by ID
  async getTicketById(ticketId: string): Promise<Ticket> {
    return apiRequest<Ticket>(`/tickets/${ticketId}`);
  },

  // Get ticket by ticket number
  async getTicketByNumber(ticketNumber: string): Promise<Ticket> {
    return apiRequest<Ticket>(`/tickets/number/${ticketNumber}`);
  },

  // Get tickets by order ID
  async getTicketsByOrderId(orderId: string): Promise<Ticket[]> {
    return apiRequest<Ticket[]>(`/tickets/order/${orderId}`);
  },

  // Get all tickets for a user
  async getTicketsByUserId(userId: string): Promise<Ticket[]> {
    return apiRequest<Ticket[]>(`/tickets/user/${userId}`);
  },

  // Cancel a ticket
  async cancelTicket(ticketId: string): Promise<string> {
    return apiRequest<string>(`/tickets/${ticketId}/cancel`, {
      method: 'POST',
    });
  },

  // Validate a ticket
  async validateTicket(qrCode: string): Promise<Ticket> {
    return apiRequest<Ticket>('/tickets/validate', {
      method: 'POST',
      body: JSON.stringify(qrCode),
    });
  },
};
